### ISSUE TITLE
Unable to compile with -DBUILD_SHARED_LIBS=1

### ISSUE THREAD

**Issue by sammy007** [monero-project/monero#1782]
tag 0.10.2

```
cmake -DBUILD_SHARED_LIBS=1 .
make
```

```
../cryptonote_basic/libcryptonote_basic.so: undefined reference to `cryptonote::BlockchainDB::get_block_from_height(unsigned long const&) const'
collect2: error: ld returned 1 exit status
src/blockchain_utilities/CMakeFiles/cn_deserialize.dir/build.make:116: recipe for target 'bin/monero-utils-deserialize' failed
make[2]: *** [bin/monero-utils-deserialize] Error 1
CMakeFiles/Makefile2:1867: recipe for target 'src/blockchain_utilities/CMakeFiles/cn_deserialize.dir/all' failed
make[1]: *** [src/blockchain_utilities/CMakeFiles/cn_deserialize.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by fluffypony** [comment]
What platform? Also did you make clean first?

**Comment by sammy007** [comment]
Oh, sorry, Linux Ubuntu 16 LTS.

**Comment by sammy007** [comment]
Clean repo clone before compilation.

**Comment by moneromooo-monero** [comment]
Try disabling LTO (-D USE_LTO=OFF in Makefile).

**Comment by sammy007** [comment]
@moneromooo-monero 

```
   if(NOT DEFINED USE_LTO_DEFAULT)
-    set(USE_LTO_DEFAULT true)
+    set(USE_LTO_DEFAULT false)
   endif()
```

also tried with `cmake -DUSE_LTO=OFF -DBUILD_SHARED_LIBS=1 .`. All the same.

**Comment by moneromooo-monero** [comment]
Do you have a function definition in src/blockchain_db/blockchain_db.cpp, line 203, like this:

block BlockchainDB::get_block_from_height(const uint64_t& height) const
{
  blobdata bd = get_block_blob_from_height(height);
  block b;
  if (!parse_and_validate_block_from_blob(bd, b))
    throw new DB_ERROR("Failed to parse block from blob retrieved from the db");

  return b;
}

**Comment by sammy007** [comment]
Sure. I am building 0.10.2 from clean clone.

```c++
block BlockchainDB::get_block_from_height(const uint64_t& height) const
{
  blobdata bd = get_block_blob_from_height(height);
  block b;
  if (!parse_and_validate_block_from_blob(bd, b))
    throw new DB_ERROR("Failed to parse block from blob retrieved from the db");

  return b;
}
```

**Comment by moneromooo-monero** [comment]
Which compiler/linker and versions ?

**Comment by sammy007** [comment]
g++ (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609
GNU ld (GNU Binutils for Ubuntu) 2.26.1

**Comment by moneromooo-monero** [comment]
Does this fix it ? https://github.com/monero-project/monero/pull/1794

**Comment by sammy007** [comment]
Nope.

```
[ 93%] Linking CXX executable ../../bin/monero-blockchain-import
../cryptonote_basic/libcryptonote_basic.so: undefined reference to `cryptonote::BlockchainDB::get_block_from_height(unsigned long const&) const'
collect2: error: ld returned 1 exit status
src/blockchain_utilities/CMakeFiles/cn_deserialize.dir/build.make:116: recipe for target 'bin/monero-utils-deserialize' failed
make[2]: *** [bin/monero-utils-deserialize] Error 1
CMakeFiles/Makefile2:1867: recipe for target 'src/blockchain_utilities/CMakeFiles/cn_deserialize.dir/all' failed
make[1]: *** [src/blockchain_utilities/CMakeFiles/cn_deserialize.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 93%] Linking CXX executable ../../bin/monero-blockchain-export
```

**Comment by hyc** [comment]
The error message is clearly complaining about undefined symbols in the shared library, so addoing blockchain_db to various executables will do nothing to address this. Apparently MacOS is like Windows and doesn't like shared libraries with undefined symbols. (By the way, that means this ought to also be broken on Windows right now.)

But again, why are we supporting building Monero internal libraries as shared libraries in the first place? You can't securely deploy an app using such shared libraries.

**Comment by sammy007** [comment]
I found this feature and using it in https://github.com/sammy007/monero-stratum to link with crypto libs and avoid copy pasting of whole `crypto` and `cryptonote_utils`, this is useful, why not keep it?

**Comment by moneromooo-monero** [comment]
It's at link time for the executable, so it seems likely that adding the lib to the executable line line will alow the compiler to find the symbol, whether required by a dependent lib, or the binary itself, no ?

**Comment by Jaqueeee** [comment]
I have the same problem on osx when building debug build. 
https://github.com/monero-project/monero/issues/1792

**Comment by sammy007** [comment]
@moneromooo-monero crossdeep fixes issue, thanks a lot.

**Comment by kenshi84** [comment]
Given Issue #1851, it's surprising to me that PR #1804 solved this issue. @sammy007 Can you confirm?

I created another [patch](https://github.com/monero-project/monero/files/829486/diff_virtual.txt) for this - could you please try this by doing
```
git apply diff_virtual.txt
```
from the project root on the latest master b67877af6f7192a302453e542c266a5cfc3182a7?

**Comment by sammy007** [comment]
It compiles with your patch on https://github.com/monero-project/monero/commit/b67877af6f7192a302453e542c266a5cfc3182a7 @kenshi84

**Comment by kenshi84** [comment]
Could you please close this issue which was solved by #1853?
